src/v/reflection/async_adl.h

```diff
@@ -60,28 +146,35 @@ struct async_adl<std::optional<T>> {
     }
 };
 
-template<typename T>
-struct async_adl<std::vector<T>> {
-    using value_type = std::remove_reference_t<std::decay_t<T>>;
+/// Specializations for most types of lists used throughout the codebase
+template<typename... Args>
+struct async_adl<std::vector<Args...>>
+  : public detail::async_adl_list<std::vector, Args...> {};
 
-    ss::future<> to(iobuf& out, std::vector<value_type> t) {
-        reflection::serialize<int32_t>(out, t.size());
-        return ss::do_with(std::move(t), [&out](auto& t) {
-            return ss::do_for_each(t, [&out](value_type& element) {
-                return async_adl<value_type>{}.to(out, std::move(element));
-            });
-        });
-    }
+template<typename... Args>
+struct async_adl<ss::circular_buffer<Args...>>
+  : public detail::async_adl_list<ss::circular_buffer, Args...> {};
+
+/// Specializations for most types of maps used throughout the codebase
+template<typename... Args>
+struct async_adl<std::map<Args...>>
+  : public detail::async_adl_map<std::map, Args...> {};
+
+template<typename... Args>
+struct async_adl<absl::btree_map<Args...>>
+  : public detail::async_adl_map<absl::btree_map, Args...> {};
+
+/// ... and hash maps
+template<typename... Args>
+struct async_adl<std::unordered_map<Args...>>
+  : public detail::async_adl_map<std::unordered_map, Args...> {};
+
+template<typename... Args>
+struct async_adl<absl::flat_hash_map<Args...>>
+  : public detail::async_adl_map<absl::flat_hash_map, Args...> {};
+
+template<typename... Args>
+struct async_adl<absl::node_hash_map<Args...>>
+  : public detail::async_adl_map<absl::node_hash_map, Args...> {};
```

Comment:
I wonder if these should go into their own files, so that we don't have to drag in every container header.